### PR TITLE
This adds an API client to handle all of the requests to the backend api endpoints

### PR DESCRIPTION
**Pull request by RizwanNiazFolio3**
### Current Status
Currently all of the components are directly using an axios instance to handle all of the api requests. The api requests also do not have any interceptors that catch errors either

### Changes/Additions
This PR adds an API client that uses axios to communicate with the backend api. The components that need to use the api can then use the Client to return an object of the desired data type needed for their functionality. This way the components are decoupled from the exact api end points and the data returned by the end points.

The PR also adds axios interceptors that can be used to automatically gain new access tokens when the old access token expires. In the case where the refresh token itself is missing or expired the user will be logged out and must then sign in again. 

It should be noted that the interceptors are currently defined for 401 errors. specific behaviors for different type of status messages can be added in a future PR

### Feedback
I'd like feedback on the PR itself, whether or not the use of the axios client and the interceptors is correct. As well as the behavior that needs to be defined for non 401 status code responses.

### CODE REVIEW

File: locallibrary-frontend/src/axios.tsx [hunk 1]
```diff
@@ -1,96 +1,228 @@
...
+	public GetAuthorDetails(id:string):Promise<AuthorDetails>{
+		return APIClient.axiosInstance.get("/catalog/api/authors/"+id)
+		.then(res=>{
+			
+			let author: AuthorDetails = {
+				first_name:null,
+				last_name:null,
+				date_of_death:null,
+				date_of_birth:null
+			}
```
Comment (shahabejaz) [inline locallibrary-frontend/src/axios.tsx:45]: Why creating an null values object first? I think you can create it with values directly.

File: locallibrary-frontend/src/axios.tsx [hunk 2]
```diff
@@ -1,96 +1,228 @@
...
+			
+			author.date_of_birth = res.data.date_of_birth 
+			author.date_of_death = res.data?.date_of_death
+			author.first_name = res.data?.first_name
+			author.last_name = res.data?.last_name
```
Comment (shahabejaz) [inline locallibrary-frontend/src/axios.tsx:50]: Can't this be created directly in one line like you have done on line 59?

File: locallibrary-frontend/src/components/pages/Authors.tsx [hunk 3]
```diff
@@ -23,16 +24,28 @@ function Authors() {
...
+         */
+        client.GetAuthorsList()
+        .then(AuthorList=>{
+            setAuthorList(AuthorList)
+        })
+        // axiosInstance.get("/catalog/api/authors")
+        // .then(
+        //     (res) => {setAuthorList(res.data)},
+        //     (error) => {console.log("There was an error retrieving author list")}//Place holder. Will be used for error handling
+        // )
```
Comment (shahabejaz) [inline locallibrary-frontend/src/components/pages/Authors.tsx:44]: Remove this code.

File: locallibrary-frontend/src/components/pages/Login.tsx [hunk 4]
```diff
@@ -14,18 +15,41 @@ function Login() {
...
+    const {isAuthenticated}: {isAuthenticated: boolean} = useContext(AuthContext)
 
     const history = useHistory()
     const [userName, setUserName]: [string, React.Dispatch<React.SetStateAction<string>>] = useState("")
     const [password, setPassword]: [string, React.Dispatch<React.SetStateAction<string>>] = useState("")
 
-    function handleSubmit(event: React.FormEvent<HTMLFormElement>){
+    useEffect(()=>{
+        if (isAuthenticated){
+            console.log("This happens")
```
Comment (shahabejaz) [inline locallibrary-frontend/src/components/pages/Login.tsx:27]: remove this console statement.

File: locallibrary-frontend/src/axios.tsx [hunk 5]
```diff
@@ -1,96 +1,228 @@
...
+		if (!APIClient.instance){
+			APIClient.instance = new APIClient(options)
+			APIClient.instance.getInterceptors()
+		}
+		return APIClient.instance
+	}
+	
+	//Used to get the detauls of an author
+	public GetAuthorDetails(id:string):Promise<AuthorDetails>{
+		return APIClient.axiosInstance.get("/catalog/api/authors/"+id)
```
[the remark this answers is not in the document]
Comment (RizwanNiazFolio3) [reply locallibrary-frontend/src/axios.tsx]: Done

File: locallibrary-frontend/src/axios.tsx (hunk 2 above)
Comment (RizwanNiazFolio3) [reply locallibrary-frontend/src/axios.tsx]: Done

**Pull request merged**